### PR TITLE
[Ray] Add elasticity to jobs launched on ray cluster

### PR DESCRIPTION
**Pull request by ljjsalt**
Elasticity - the execution of placement groups are pending tasks that will be scheduled by GCS when resources become available.

Related PR: #572 

Test plan:

Mock cluster scaling with `ray.cluster_utils`.

### CODE REVIEW

File: torchx/schedulers/ray/ray_driver.py [hunk 1]
```diff
@@ -77,77 +102,189 @@ def load_actor_json(filename: str) -> List[RayActor]:
...
+            True  # need more actors before any actor finishes and exits
         )
-    return pg
+        self.command_actors_count: int = 0  # number of created command actors
 
+    def init_placement_groups(self) -> None:
+        """Initialize all placement groups needed for this job"""
+        # find the actor specifications of a given placement group
+        replica_ix_of_pg: List[int] = [0] + list(
+            range(self.min_nnodes, self.max_nnodes + 1)  # pyre-ignore
```
Comment (d4l3k) [inline torchx/schedulers/ray/ray_driver.py:152]: why is there a pyre ignore?

should at least be a pyre-fixme with the error message

File: torchx/schedulers/ray/ray_driver.py [hunk 2]
```diff
@@ -77,77 +102,189 @@ def load_actor_json(filename: str) -> List[RayActor]:
...
+                        )
+                    else:
+                        self.active_tasks.append(
+                            actor.exec_module.remote(
+                                self.rank_0_address, self.rank_0_port, result.id
+                            )
+                        )
+                    self.command_actors_count += 1
+            elif isinstance(result, TaskCompleted):
+                self.need_more_actors = False  # don't need more actors
```
Comment (d4l3k) [inline torchx/schedulers/ray/ray_driver.py:249]: maybe should name this terminating? and then we can do `if not self.terminating` above

File: torchx/schedulers/ray/ray_driver.py [hunk 3]
```diff
@@ -77,77 +102,189 @@ def load_actor_json(filename: str) -> List[RayActor]:
...
+        """Creating all command actors in all placement groups"""
+        # find the placement group index for a replica(actor's specification)
+        pg_ix_of_replica: List[int] = [
+            max(0, i - self.min_nnodes + 1)  # pyre-ignore
+            for i in range(len(self.replicas))
+        ]
+        # create the actors
+        for i in range(len(self.replicas)):
+            pg_ix = pg_ix_of_replica[i]
+            pg = self.placement_groups[pg_ix]  # find the created placement group
```
Comment (d4l3k) [inline torchx/schedulers/ray/ray_driver.py:212]: I'm assuming if the placement_group hasn't started this just queues the actor?

File: torchx/schedulers/ray/ray_driver.py [hunk 4]
```diff
@@ -77,77 +102,189 @@ def load_actor_json(filename: str) -> List[RayActor]:
...
+        return False
+
+    def run(self) -> None:
+        """This is the main loop the ray driver, it executes the user script on the scheduled nodes,
+        and restart the failed nodes(node failures). The loop ends when all the actors that joining
+        the job exits."""
+        self.need_more_actors = True
+        self.command_actors_count = 0
+        # Await return result of remote ray function and initialize new command actors
+        while len(self.active_tasks) > 0:
```
Comment (d4l3k) [inline torchx/schedulers/ray/ray_driver.py:271]: wow this is much cleaner :)

**Review by d4l3k — COMMENTED** [review]
this is looking much cleaner, just a few nits and then I think we can merge this :) thanks for polishing this up! :rocket:

File: torchx/schedulers/ray/ray_driver.py (hunk 3 above)
Comment (ljjsalt) [reply torchx/schedulers/ray/ray_driver.py:212]: yes, it's just telling the ray cluster that I want to put a command actor in this placement group. And only after the placement group has been scheduled to an available node, the actors get created.

File: torchx/specs/api.py [hunk 5]
```diff
@@ -321,6 +321,7 @@ class Role:
 
     name: str
     image: str
+    min_nnodes: Optional[int] = None
```
Comment (d4l3k) [inline torchx/specs/api.py:324]: min_replicas

**Review by d4l3k — COMMENTED** [review]
just one more change :)

File: torchx/specs/api.py (hunk 5 above)
Comment (ljjsalt) [reply torchx/specs/api.py]: fixed

**Review by d4l3k — APPROVED** [review]

**Review by priyaramani — APPROVED** [review]

File: torchx/schedulers/test/ray_scheduler_test.py [hunk 6]
```diff
@@ -322,44 +324,60 @@ def test_list_throws_without_address(self) -> None:
...
-        def decrement_reference(cls) -> None:
-            cls.reference_count = cls.reference_count - 1
-            if cls.reference_count == 0:
-                cls.teardown_ray_cluster()
+        @property
+        def workers(self) -> List[ray.node.Node]:
+            return list(self._cluster.worker_nodes)
+
+        def add_node(self, num_cpus: int = 1) -> None:
+            # add 1 node with 2 cpus to the cluster
```
[the remark this answers is not in the document]
Comment (ljjsalt) [reply torchx/schedulers/test/ray_scheduler_test.py]: yes，n cpus, fixed

**Pull request closed without merging**